data_structures/tree_234.cpp

```diff
@@ -0,0 +1,1085 @@
+                            node->GetItemRightChild(index + 1);
+                        if (right_sibling->Is34Node()) {
+                            LeftRotate(node, index + 1);
+                            break;
+                        }
+                    }
+
+                    // we do a merge. We avoid merging the target item, which
+                    // may trigger another merge in the recursion process.
+                    if (index > 0) {
+                        Merge(node, index - 1);
+                        break;
+                    }
+
+                    Merge(node, index + 1);
+
+                } while (false);
+            }
+
+            // refresh the left_child and right_child since they may be invalid
+            // because of merge
+            left_child = node->GetItemLeftChild(index);
+            right_child = node->GetItemRightChild(index);
+
+            if (left_child->Is34Node()) {
+                int predecessor_item = GetTreeMaxItem(left_child);
+                node->SetItem(node->GetItemIndex(item), predecessor_item);
+
+                node = left_child;
+                item = predecessor_item;
+                continue;
+            }
+
+            if (right_child->Is34Node()) {
+                int successor_item = GetTreeMinItem(right_child);
+                node->SetItem(node->GetItemIndex(item), successor_item);
+                node = right_child;
+                item = successor_item;
+                continue;
+            }
+        }
+
+        Node *next_node = node->GetNextPossibleChild(item);
+
+        if (next_node->Is34Node()) {
+            node = next_node;
+            continue;
+        }
+
+        if (TryRightRotate(node, next_node)) {
+            node = next_node;
+            continue;
+        }
+
+        if (TryLeftRotate(node, next_node)) {
+            node = next_node;
+            continue;
+        }
+
+        // get here means both left sibling and right sibling of next_node is
+        // 2-node, so we do merge
+        int child_index = node->GetChildIndex(next_node);
+        if (child_index > 0) {
+            node = Merge(node, child_index - 1);
+        } else {
+            node = Merge(node, child_index);
+        }
+
+    }  // while
+
+    return false;
+}
+
+int Tree234::GetTreeMaxItem(Node *tree) {
+    assert(tree);
+    int max = 0;
+
+    while (tree) {
+        max = tree->GetMaxItem();
+        tree = tree->GetRightmostChild();
+    }
+
+    return max;
+}
+
+int Tree234::GetTreeMinItem(Node *tree) {
+    assert(tree);
+    int min = 0;
+
+    while (tree) {
+        min = tree->GetMinItem();
+        tree = tree->GetLeftmostChild();
+    }
+
+    return min;
+}
+
+void Tree234::Print(const char *file_name) {
+    if (!file_name) {
+        file_name = "out.dot";
+    }
+
+    FILE *f = fopen(file_name, "wb");
+    fprintf(f, "digraph G {\n");
+    fprintf(f, "node [shape=record]\n");
+
+    int index = 0;
+
+    struct NodeInfo {
```

Comment:
Add a brief description of the structure here (in the code).
```suggestion
    /**
     * @brief description of the struct
     */
    struct NodeInfo {
```